### DISCUSSION TITLE
Could BlueRetro via Wii EXT emulate a Wii U Pro Controller?

### DISCUSSION THREAD

**Discussion by vaguerant** [darthcloud/BlueRetro#825] [Ideas]
Just to be clear, I know a real Wii U Pro Controller can be paired with a BlueRetro adapter, but I'm asking about a fake WUPC going into Wii EXT on the wired side, not a real one on the wireless Bluetooth side.

My understanding is that internally, the Wii U Pro Controller behaves very similarly to a Classic Controller, in the sense of being a "Wiimote" with a Classic Controller "in" the EXT port. While it's not physically a Wiimote and dongle, from the console (Wii U) side, that is how it is read.

Based on the way it's used in Wii U homebrew, it looks like it is basically identical to a Classic Controller but with slightly longer reports. See [libogc2](https://github.com/extremscorner/libogc2/blob/master/wiiuse/classic.c) for example, where they are treated as `cc->type` `0` (Wii Classic Controller), `1` (Wii Classic Controller Pro) or `2` (Wii U Pro Controller) based on the report length.

Would it be possible for the BlueRetro adapter to report itself to the Wiimote as a Wii U Pro Controller such that a Wii U would detect that Wii Remote and Wii EXT BlueRetro combo as a real WUPC?

The practical use of this over identifying as a normal Wii Classic Controller/Pro is that it would allow BlueRetro to control Wii U software which supports the Pro Controller but **not** the Classic Controller, including some of its most popular games like New Super Mario Bros. U, Donkey Kong Country: Tropical Freeze, Zelda: Wind Waker HD, Twilight Princess HD and Breath of the Wild.

**Comment by darthcloud** [comment]
Yes that would be possible and straight forward to do.

But the BT name would be the wiimote one rather than the Wiiu pro one.

So for this to work things needs to ignore that at least.

**Reply by vaguerant** [reply]
I asked GaryOderNichts about this, as they wrote [Bloopair](https://github.com/GaryOderNichts/Bloopair), a homebrew tool which modifies the Wii U's controller library in order to have other controllers be recognized as WUPCs. Unfortunately, the Wii U **does** use the Bluetooth name to identify Wii Remotes vs. Wii U Pro Controllers, so this would not work out of the box on unmodified consoles.

(In case you're wondering, Bloopair and the Wii U itself do not support Bluetooth LE, so Bloopair cannot be used to connect e.g. an Xbox Series controller directly, which is where this idea of going via BlueRetro and a Wii Remote comes from in the first place.)

In theory, the Bluetooth device name issue is probably not insurmountable; homebrew developers have essentially complete control over the Wii U at this point, so it's likely that too could be changed. However, Gary also notes:

>  I don't think just telling the system that a Wiimote has a "Pro Controller Extension" would work, since there is still some special stuff going on when it comes to pro controllers.

It sounds like it would require considerable effort on the Wii U side at least to make this happen. This is a pretty niche idea in the first place; it would require people to have a very specific set of hardware (a modified Wii U, some Bloopair-unsupported controller/s, a BlueRetro adapter with Wii EXT out and a Wii Remote), so I doubt there's really the impetus on either end to make this happen.

Personally, between BlueRetro and Bloopair all of my controllers and consoles are covered, so even I don't actually need this feature, it just seems like such an interesting novelty because of how odd the Wii U Pro Controller is. I was hopeful that it would all come together and "just work", but sadly that doesn't look like it's the case.

**Comment by Invictaz** [comment]
If Blueretro had USB output then the route via the wiimote isn't needed at all. Just spoof a wired wiiu pro controller over USB, just like the Brook adapter that was sold a few years ago.
